Unrevert "[Sema] Setter has incorrect mutating-ness inside class-constrained protocol extension"

theblixguy opened the pull request:
This un-reverts #26695. The Core Team accepts this as a [bug fix](https://forums.swift.org/t/class-constrained-protocol-extension-has-incorrect-mutatingness/27962/11).

Resolves [SR-11298](https://bugs.swift.org/browse/SR-11298).

theblixguy committed:
Revert "Revert "[Sema] Setter has incorrect mutating-ness inside class-constrained protocol extension""

This reverts commit edb153afd1d71fb869c4add80d0b47451092d62b.

theblixguy committed:
[Test] Update tests

theblixguy committed:
[Changelog] Add a changelog entry

jrose-apple commented on CHANGELOG.md:
````diff
@@ -26,9 +26,32 @@ CHANGELOG
+      set { someProperty = newValue }
+    }
+  }
+  ```
````
Can you include the example of the code that would break?

jrose-apple replied:
And maybe how to work around that.

theblixguy replied:
Sure, added one now.

theblixguy committed:
[Changelog] Add an example for the broken case

theblixguy committed:
Merge branch 'master' into unrevert/SR-11298

jrose-apple commented on CHANGELOG.md:
```diff
@@ -48,6 +48,33 @@ Swift Next
+    
+    var anotherProperty2: Int {
+      get { return someProperty }
+      mutating set { someProperty = newValue } // Okay
```
Wait, so the default changes to `nonmutating` but you're still allowed to *write* `mutating`? That's confusing but also a very easy answer. Can you add this as a test case to make sure it works?

theblixguy replied:
There's already a test case (see the last one). Do you mean a test case where I have explicitly written `mutating`?

theblixguy replied:
Actually, I _assumed_ we can add `mutating` to the setter (by looking at the fix-it), but I didn't actually try it. 

The fix-it is actually bogus because adding `mutating` will not make the error go away. What actually needs to be done is marking the protocol requirement as `nonmutating` as well, so we need to either not offer the fix-it or tweak it. I suppose I can do that in a follow-up PR.

jrose-apple replied:
I was thinking the workaround is "assign into a temporary var, even though the change won't be reflected elsewhere". We definitely shouldn't offer the fix-it if it doesn't work. Changing the requirement probably isn't acceptable.

theblixguy replied:
Yeah, I've removed the fix-it now for that case.

theblixguy committed:
[Diagnostics] Do not offer a mutating fix-it if we have a mutating protocol requirement

Do not offer a mutating fix-it if we have a mutating protocol requirement and we're assigning to it from an implicitly nonmutating setter inside a protocol extension

theblixguy committed:
[Changelog] Update changelog

jrose-apple commented on CHANGELOG.md:
````diff
@@ -26,6 +26,53 @@ CHANGELOG
+  }
+  ```
+
+  To resolve this, explicitly mark the setter as `nonmutating` in the protocol.
````
I still think this is the wrong recommendation: the extension is constrained because the protocol is not class-bound itself, and that means the mutating-ness of the existing property might be relevant.

theblixguy replied:
Sure, I suppose we can just drop this and just keep the non-working example. I thought it would make sense to offer at least one possible solution (perhaps this could be reworded), but it's probably okay to not offer one.

jrose-apple replied:
No, I mean, the right answer is

```swift
var anotherProperty1: Int {
  get { return someProperty }
  set {
    var mutableSelf = self
    mutableSelf.someProperty = newValue
    assert(mutableSelf === self, "assigning the property really did reset 'self'")
  }
}
```

…though I'm not sure if we want to explain that to people.

theblixguy replied:
I think it's worth adding a solution.

theblixguy replied:
Added it now.

theblixguy committed:
[Changelog] Add workaround for nonmutating setter

jrose-apple commented on CHANGELOG.md:
````diff
@@ -66,12 +66,22 @@ Swift Next
   ```
 
-  To resolve this, explicitly mark the setter as `nonmutating` in the protocol.
+  **Workaround**: Define a new mutable property inside the setter that has a reference to `self`:
````
Nitpick: "mutable variable" is probably easier for people to understand.

theblixguy replied:
Fixed!

jrose-apple reviewed (commented):
Thanks!

theblixguy committed:
[Changelog] Change 'mutable property' to 'mutable variable'

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5703,12 +5704,31 @@ void VarDecl::emitLetToVarNoteIfSimple(DeclContext *UseDC) const {
+
+        // Do not suggest the fix-it if we have an implicitly nonmutating
+        // setter in a protocol extension and we're assigning to a mutating
+        // protocol requirement.
```
This is very specific but still feels like it's not getting at the interesting bit. How about just "if the `Self` type (`AD->getDeclContext()->getSelfInterfaceType()`) is class-bound (`isAnyClassReferenceType()`)?"

theblixguy replied:
Sure, do you mean we should replace all of the checks below with a simple "is Self type class-bound or not"?

jrose-apple replied:
Yeah. If Self is class-bound, you can't insert `mutating`, at least not today, and that covers all the cases we care about introduced by your other change.

theblixguy replied:
Alright, done!

theblixguy committed:
[Typechecker] Use a 'is Self class-bound?' check instead

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1957,8 +1957,17 @@ void TypeChecker::validateDecl(OperatorDecl *OD) {
+      extensionRequiresClass =
+          ED->getGenericSignature()->requiresClass(ED->getSelfInterfaceType());
+    }
+    return !contextTy->hasReferenceSemantics() && !extensionRequiresClass;
```
I'm just noticing this now, sorry, but this will do all the work even if the immediate context type has reference semantics and it wasn't important to check the extension. Please use early returns.

theblixguy replied:
I think we should also be using `getSelfTypeInContext()` here as well, right? Then we won’t need the extension check!

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5703,12 +5703,19 @@ void VarDecl::emitLetToVarNoteIfSimple(DeclContext *UseDC) const {
+        // Do not suggest the fix-it if `Self` is a class type.
+        if (AD->getDeclContext()
+                ->getSelfTypeInContext()
+                ->isAnyClassReferenceType()) {
```
In retrospect, should this just be `!doesContextHaveValueSemantics(AD->getDeclContext())` instead of what I said? (And whoops on `getSelfInterfaceType` vs. `getSelfTypeInContext`, yours is correct.)

theblixguy replied:
Hmm `doesContextHaveValueSemantics` is not accessible here though and it also calls `isReferenceTypeImpl()`, which is what `isAnyClassReferenceType()` and `hasReferenceSemantics()` calls (difference between the two that the former doesn't include functions but the latter does).

I think we don't need the `doesContextHaveValueSemantics` check as `getSelfTypeInContext()` will return the mapped `Self` type of the protocol or extension (it would be a `PrimaryArchetypeType`) and then the `isReferenceTypeImpl()` check would look into it to check if it requires class.

What do you think?

theblixguy replied:
Maybe we can move `doesContextHaveValueSemantics()` into `DeclContext`.

jrose-apple replied:
I think `getSelfTypeInContext()` is going to be more expensive than checking against the generic signature (because the former would actually have to form the type), but either way is okay, as is moving the predicate into DeclContext. @slavapestov, any opinions?

theblixguy replied:
I've moved `doesContextHaveValueSemantics()` into `DeclContext` and switched to `getSelfTypeInContext()`, which is simpler to use, but I can revert back to the generic signature check if it's less expensive.

jrose-apple commented:
Thanks for sticking with this, by the way. Sorry for only noticing one or two things each time—I keep thinking "oh, this is an unrevert and it was basically ready to go in last time so I don't really need to look at it closely".

theblixguy committed:
[AST] Adds 'hasValueSemantics()' to DeclContext and uses 'getSelfTypeInContext()' instead

theblixguy commented:
Running tests while I wait for approval:

@swift-ci please smoke test

theblixguy commented:
@swift-ci please test source compatibility

theblixguy committed:
[AST] hasValueSemantics() should check whether the current context is a type context before proceeding

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please test source compatibility

jrose-apple commented on lib/AST/DeclContext.cpp:
```diff
@@ -1025,6 +1025,9 @@ DeclContextKind DeclContext::getContextKind() const {
 
 bool DeclContext::hasValueSemantics() const {
+	if (!isTypeContext())
+		return false;
```
Looks like some tabs snuck in.

theblixguy replied:
Whoops! I’ll fix it after the test finishes.

jrose-apple replied:
Still not fixed!

jrose-apple replied:
Though I'm not sure I like the idea of testing it here, rather than asserting it and having callers check.

theblixguy replied:
`getSelfTypeInContext` already asserts that, but seems like this check is called from places where we don’t have a type context and thus end up firing the assert (ex: when building the stdlib). 

The code before was calling `getDeclaredInterfaceType` which would just return an `ErrorType` or null if we couldn’t get the type.

jrose-apple replied:
> `getSelfTypeInContext` already asserts that, but seems like this check is called from places where we don’t have a type context and thus end up firing the assert (ex: when building the stdlib).

Yeah, I remember seeing you add it in here. But asking `hasValueSemantics` doesn't actually make sense for a non-type context, and I don't think we should silently accept it by saying "no".

theblixguy replied:
Alright, I have removed the check.

theblixguy commented:
Alright, all tests have passed :shipit:

theblixguy commented:
Is this good to merge now? Just need an approval before I do that (and fix the indentation).

theblixguy committed:
[AST] Fix indenation in hasValueSemantics()

theblixguy committed:
[AST] Remove isTypeContext() check from hasValueSemantics() and move the responsibility to the caller

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1959,7 +1959,9 @@ void TypeChecker::validateDecl(OperatorDecl *OD) {
-    if (!FD->isInstanceMember() || !FD->getDeclContext()->hasValueSemantics()) {
+    auto functionDC = FD->getDeclContext();
+    if (!FD->isInstanceMember() ||
+        (functionDC->isTypeContext() && !functionDC->hasValueSemantics())) {
```
`FD->isInstanceMember()` already implies `functionDC->isTypeContext()`, no? (here and below)

theblixguy replied:
Hmm I am not sure, it doesn't check for a type context internally (but for a context name and the index of the Self parameter). Let me verify

jrose-apple replied:
That's the `isInstanceMember` in Parser. The one in ValueDecl does start off by checking for a type context.

theblixguy replied:
Oops, you're right. Xcode took me to the wrong place and I didn't notice. I'll remove the extra check now.

theblixguy replied:
Done.

theblixguy committed:
[Typechecker] Remove a few unnecessary isTypeContext() checks as its implifed by isInstanceMember()

jrose-apple commented:
Thanks again for sticking with it!

theblixguy commented:
@swift-ci please test

theblixguy commented:
@swift-ci please test source compatibility
